great_expectations/experimental/datasources/experimental_base_model.py

```diff
@@ -40,6 +40,12 @@ class ExperimentalBaseModel(pydantic.BaseModel):
     class Config:
         extra = pydantic.Extra.forbid
 
+    def __key(self):
+        return tuple(self.dict())
+
+    def __hash__(self):
+        return hash(self.__key())
```

Comment:
Is this because of `_get_execution_engine` lru cache?
Which models specifically get can be passed to `_get_execution_engine()` we might not want to do the on the `ExperimentalBaseModel`.

Reply:
It looks like instances of custom classes are normally hashable by default.
But `pydantic` models are not (by default).

https://docs.python.org/3.8/glossary.html (search `hashable`)

>An object is hashable if it has a hash value which never changes during its lifetime (it needs a [__hash__()](https://docs.python.org/3.8/reference/datamodel.html#object.__hash__) method), and can be compared to other objects (it needs an [__eq__()](https://docs.python.org/3.8/reference/datamodel.html#object.__eq__) method). Hashable objects which compare equal must have the same hash value.

>...

>Most of Python’s immutable built-in objects are hashable; mutable containers (such as lists or dictionaries) are not; immutable containers (such as tuples and frozensets) are only hashable if their elements are hashable. Objects which are instances of user-defined classes are hashable by default. They all compare unequal (except with themselves), and their hash value is derived from their [id()](https://docs.python.org/3.8/library/functions.html#id).

https://docs.pydantic.dev/usage/model_config/#options (see `frozen`)

Reply:
Yes, this is due to `lru_cache` needing all args to be hashable. We can move this down to the datasource level if we chose to do so. For some reason the `lru_cache` also needs the `Datasource` to be hashable or it throws a `TypeError`.

Reply:
If this is not a property you need, you could manually cache it by setting a variable on the data source, eg `my_datasource._execution_engine` which get's populated the first time `execution_engine` is called. You could store the kwargs or other metadata if you need to know when to reconnect.